### PR TITLE
Ban decorator usage of dynamo_timed

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132328

This is a more manual version of https://github.com/pytorch/pytorch/pull/132073 that just manually creates the new function at each call site instead of magicking it with clone. Review with whitespace diffs off.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -543,17 +543,23 @@ def maybe_realign_inputs(
...
 @time_and_log(attr="compilation time (in seconds)")
 # Need this decorator for compile_fx_inner even if we already have one for
 # compile_fx. The reason is the compilation for backward graph may happen after
 # compile_fx return and we may want to use the _LazyGraphModule for compiling
 # the backward graph as well.
-@_use_lazy_graph_module(dynamo_config.use_lazy_graph_module)
 @with_fresh_cache_if_config
-@dynamo_utils.dynamo_timed(phase_name="inductor_compile", fwd_only=False)
-def compile_fx_inner(
+def _compile_fx_inner(
```
Comment (ezyang) [inline torch/_inductor/compile_fx.py:562]: These decorators are temporarily reordered, but I don't think it matters. Once all of these become context managers I can restore original order.

File: torch/_inductor/compile_fx.py [hunk 2]
```diff
@@ -1085,7 +1090,9 @@ def cudagraphify(
     def run(new_inputs):
         nonlocal compiled_fn
         if compiled_fn is None:
-            with dynamo_utils.preserve_rng_state():
+            with dynamo_utils.dynamo_timed(
+                "cudagraphify"
+            ), dynamo_utils.preserve_rng_state():
```
Comment (ezyang) [inline torch/_inductor/compile_fx.py:1095]: This is a bug fix, the original one didn't capture the right logic

File: torch/fx/experimental/validator.py [hunk 3]
```diff
@@ -517,8 +518,10 @@ def add_assertion(self, e: Union[z3.BoolRef, sympy.Basic]) -> None:
             self._assertions.add(ref)
 
         def validate(self) -> None:
-            from torch._dynamo.utils import dynamo_timed
+            with dynamo_timed("TranslationValidator.validate"):
+                return self.validate()
```
Comment (ezyang) [inline torch/fx/experimental/validator.py:522]: I'm not sure this ever worked lol

File: .lintrunner.toml [hunk 4]
```diff
@@ -940,6 +940,25 @@ command = [
...
 
+[[linter]]
+code = 'CONTEXT_DECORATOR'
+include_patterns = [
+    'torch/**',
+]
+command = [
+    'python3',
+    'tools/linter/adapters/grep_linter.py',
+    '--pattern=@.+(dynamo_timed)',
```
Comment (albanD) [inline .lintrunner.toml:951]:
```suggestion
    '--pattern=@.*(dynamo_timed)',
```

Or you specifically don't want to capture plain `@dynamo_timed` ?

**Review by albanD — APPROVED** [review]
Small question on lint but sounds good.
Not as nice as the copy but most likely a lot more reliable!

File: .lintrunner.toml (hunk 4 above)
Comment (ezyang) [reply .lintrunner.toml:951]: I want to catch @torch._dynamo.utils.dynamo_timed() too

Comment (albanD) [reply .lintrunner.toml:951]: Right, so you want the `*` version, not `+` ?

Comment (ezyang) [reply .lintrunner.toml:951]: oh whoops you right

**Comment by clee2000** [comment]
Sorry my bad it was actually the commit right before this one

**Comment by bdhirsh** [comment]
very excited for this PR!! (reading CProfile graphs previously was pretty annoying)

**Pull request closed without merging**